### PR TITLE
fix: check resource namespaces are managed

### PR DESCRIPTION
**Pull request by jopit**
Signed-off-by: John Pitman <jpitman@redhat.com>

Related to https://github.com/argoproj/argo-cd/issues/4329

Return an error from `clusterCache.GetManagedLiveObjs` if we find a target object in a namespace that is not managed by the cluster.

### CODE REVIEW

File: pkg/cache/cluster.go [hunk 1]
```diff
@@ -661,13 +661,34 @@ func (c *clusterCache) IsNamespaced(gk schema.GroupKind) (bool, error) {
 	return false, errors.NewNotFound(schema.GroupResource{Group: gk.Group}, "")
 }
 
+func (c *clusterCache) managesNamespace(namespace string) bool {
+	if len(c.namespaces) == 0 {
+		return true
+	}
+	if len(namespace) == 0 {
+		return true
```
Comment (alexmt) [inline pkg/cache/cluster.go:669]: I think we should return false here. If `len(namespace) == 0` is true then resource is a cluster-level resource. Cluster level resources are not managed if cluster is connected in namespaced mode so we should return false.

```golang
if namespace == "" {
	return true
}
```

**Review by alexmt — CHANGES_REQUESTED** [review]

File: pkg/cache/cluster.go (hunk 1 above)
Comment (jopit) [reply pkg/cache/cluster.go]: I've made the suggested change.

**Comment by jopit** [comment]
Hi @alexmt the SonarCloud failure look like it's due to duplicated code in the tests I added.  I think that eliminating the duplication will make the tests harder to understand.  Do you want me to eliminate the duplication anyway?

**Review by alexmt — APPROVED** [review]
LGTM

**Comment by alexmt** [comment]
@jopit the SonarCloud check is optional. I think it is ok to have duplication in test

**Pull request merged**